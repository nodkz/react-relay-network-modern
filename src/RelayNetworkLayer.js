/* @flow */

import { Network } from 'relay-runtime';
import RelayRequest from './RelayRequest';
import fetchWithMiddleware from './fetchWithMiddleware';
import type {
  Middleware,
  RawMiddleware,
  MiddlewareSync,
  FetchFunction,
  FetchHookFunction,
  SubscribeFunction,
  RNLExecuteFunction,
} from './definition';

type RelayNetworkLayerOpts = {|
  subscribeFn?: SubscribeFunction,
  beforeFetch?: FetchHookFunction,
  noThrow?: boolean,
|};

export default class RelayNetworkLayer {
  _middlewares: Middleware[];
  _rawMiddlewares: RawMiddleware[];
  _middlewaresSync: RNLExecuteFunction[];
  execute: RNLExecuteFunction;
  +fetchFn: FetchFunction;
  +subscribeFn: ?SubscribeFunction;
  +noThrow: boolean;

  constructor(
    middlewares: Array<?Middleware | MiddlewareSync>,
    rawMiddlewares?: Array<?RawMiddleware>,
    opts?: RelayNetworkLayerOpts
  ) {
    this._middlewares = [];
    this._rawMiddlewares = [];
    this._middlewaresSync = [];
    this.noThrow = false;

    const mws = Array.isArray(middlewares) ? (middlewares: any) : [middlewares];
    mws.forEach(mw => {
      if (mw) {
        if (mw.execute) {
          this._middlewaresSync.push(mw.execute);
        } else {
          this._middlewares.push(mw);
        }
      }
    });

    const rmws = Array.isArray(rawMiddlewares) ? (rawMiddlewares: any) : [rawMiddlewares];
    rmws.forEach(rmw => {
      if (rmw) {
        this._rawMiddlewares.push(rmw);
      }
    });

    if (opts) {
      this.subscribeFn = opts.subscribeFn;
      this.noThrow = opts.noThrow === true;

      // TODO deprecate
      if (opts.beforeFetch) {
        this._middlewaresSync.push((opts.beforeFetch: any));
      }
    }

    this.fetchFn = (operation, variables, cacheConfig, uploadables) => {
      for (let i = 0; i < this._middlewaresSync.length; i++) {
        const res = this._middlewaresSync[i](operation, variables, cacheConfig, uploadables);
        if (res) return res;
      }

      const req = new RelayRequest(operation, variables, cacheConfig, uploadables);
      return fetchWithMiddleware(req, this._middlewares, this._rawMiddlewares, this.noThrow);
    };

    const network = Network.create(this.fetchFn, this.subscribeFn);
    this.execute = network.execute;
  }
}
