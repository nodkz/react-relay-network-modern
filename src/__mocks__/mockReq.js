/* @flow */
/* eslint-disable import/prefer-default-export, no-param-reassign */

import type RelayNetworkLayer from '../RelayNetworkLayer';
import type RelayResponse from '../RelayResponse';

type ReqData = {
  query?: string,
  variables?: Object,
  files?: any,
};

type ReqId = string;

class MockReq {
  reqid: ReqId;
  reqData: ReqData;
  error: Error;
  payload: Object;

  constructor(reqid?: ReqId, reqData?: ReqData = {}) {
    this.reqid = reqid || Math.random().toString();
    this.reqData = reqData;
  }

  getID(): ReqId {
    return this.reqid;
  }

  getQueryString(): string {
    return this.reqData.query || '';
  }

  getDebugName(): string {
    return `debugname${this.reqid}`;
  }

  getVariables(): Object {
    return this.reqData.variables || {};
  }

  getFiles(): any {
    return this.reqData.files;
  }

  reject(err: Error) {
    this.error = err;
  }

  resolve(resp: Object) {
    this.payload = resp;
  }

  execute(rnl: RelayNetworkLayer): Promise<RelayResponse> {
    const operation = ({
      id: this.getID(),
      text: this.getQueryString() || '',
    }: any);
    const variables = this.getVariables() || {};
    const cacheConfig = {};
    const uploadables = this.getFiles();

    const res = (rnl.fetchFn(operation, variables, cacheConfig, uploadables): any);

    const promise = new Promise((resolve, reject) => {
      res.subscribe({
        complete: () => {},
        error: (error) => reject(error),
        next: (value) => resolve(value),
      });
    });

    // avoid unhandled rejection in tests
    promise.catch(() => {});

    // but allow to read rejected response
    return promise;
  }
}

export function mockReq(reqid?: ReqId | number, data?: ReqData): MockReq {
  return new MockReq(reqid ? reqid.toString() : undefined, data);
}

export function mockMutationReq(reqid?: ReqId | number, data?: ReqData): MockReq {
  return new MockReq(reqid ? reqid.toString() : undefined, {
    query: 'mutation {}',
    ...data,
  });
}

export function mockFormDataReq(reqid?: ReqId | number, data?: ReqData): MockReq {
  return new MockReq(reqid ? reqid.toString() : undefined, {
    files: { file1: 'data' },
    ...data,
  });
}

export function mockReqWithSize(reqid: ReqId | number, size: number): MockReq {
  return mockReq(reqid, { query: `{${'x'.repeat(size)}}` });
}

export function mockReqWithFiles(reqid: ReqId | number): MockReq {
  return mockReq(reqid, { files: { file1: 'data', file2: 'data' } });
}
