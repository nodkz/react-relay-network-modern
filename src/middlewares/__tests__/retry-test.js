/* @flow */

import fetchMock from 'fetch-mock';
import RelayNetworkLayer from '../../RelayNetworkLayer';
import { mockReq } from '../../__mocks__/mockReq';
import retryMiddleware, { delayExecution, promiseWithTimeout } from '../retry';

async function sleep(timeout: number): Promise<void> {
  return new Promise(resolve => {
    setTimeout(resolve, timeout);
  });
}

describe('middlewares/retry', () => {
  describe('promiseWithTimeout()', () => {
    it('should return Promise result if not reach timeout ', async () => {
      const p = Promise.resolve(5);
      const r = await promiseWithTimeout(p, 1000, () => Promise.resolve(0));
      expect(r).toBe(5);
    });

    it('should run `onTimeout` when timout is reached', async () => {
      const p = new Promise(resolve => {
        setTimeout(() => {
          resolve(333);
        }, 20);
      });
      const onTimeout = jest.fn(() => {
        return Promise.resolve(555);
      });
      const r = await promiseWithTimeout(p, 10, onTimeout);
      expect(onTimeout).toHaveBeenCalledTimes(1);
      expect(r).toBe(555);
    });
  });

  describe('delayExecution()', () => {
    it('should run function after delay', async () => {
      const execFn = jest.fn(() => Promise.resolve(777));
      const forceRetryWhenDelay = jest.fn();

      const promise = delayExecution(execFn, 10, forceRetryWhenDelay);
      await sleep(5);
      expect(execFn).toHaveBeenCalledTimes(0);
      await sleep(10);
      expect(execFn).toHaveBeenCalledTimes(1);

      const r = await promise;
      expect(r).toBe(777);
    });

    it('should run function after call `runNow`', async () => {
      const execFn = jest.fn(() => Promise.resolve(888));
      const forceRetryWhenDelay = jest.fn(cb => {
        setTimeout(() => cb(), 10);
      });

      const promise = delayExecution(execFn, 1000, forceRetryWhenDelay);
      await sleep(5);
      expect(forceRetryWhenDelay).toHaveBeenCalledTimes(1);
      expect(execFn).toHaveBeenCalledTimes(0);
      await sleep(10);
      expect(execFn).toHaveBeenCalledTimes(1);

      const r = await promise;
      expect(r).toBe(888);
    });
  });

  describe('middleware', () => {
    beforeEach(() => {
      fetchMock.restore();
    });

    it('should make retries', async () => {
      // First 2 requests return code 500,
      // 3rd request returns code 200
      let attempt = 0;
      fetchMock.mock({
        matcher: '/graphql',
        response: () => {
          attempt++;
          if (attempt < 3) {
            return { status: 500, body: '' };
          }
          return {
            status: 200,
            body: { data: 'PAYLOAD' },
          };
        },
        method: 'POST',
      });
      const rnl = new RelayNetworkLayer([
        retryMiddleware({
          retryDelays: () => 1,
          logger: false,
        }),
      ]);

      const res = await mockReq(1).execute(rnl);
      expect(res.data).toEqual('PAYLOAD');

      const reqs = fetchMock.calls('/graphql');
      expect(reqs).toHaveLength(3);
      expect(reqs).toMatchSnapshot();
    });

    it('should retry request on timeout', async () => {
      let attempt = 0;

      // First 2 requests answered after 50ms
      // 3rd request returns without delay
      fetchMock.mock({
        matcher: '/graphql',
        response: () => {
          attempt++;
          return new Promise(resolve => {
            setTimeout(
              () =>
                resolve({
                  status: 200,
                  body: { data: 'PAYLOAD' },
                }),
              attempt <= 2 ? 50 : 0
            );
          });
        },
        method: 'POST',
      });
      const rnl = new RelayNetworkLayer([
        retryMiddleware({
          fetchTimeout: 20,
          retryDelays: () => 1,
          logger: false,
        }),
      ]);
      await mockReq(1).execute(rnl);
      const reqs = fetchMock.calls('/graphql');
      expect(reqs).toHaveLength(3);
      expect(reqs).toMatchSnapshot();
    });

    it('should throw error on timeout', async () => {
      // returns request after 100ms
      fetchMock.mock({
        matcher: '/graphql',
        response: () => {
          return new Promise(resolve => {
            setTimeout(
              () =>
                resolve({
                  status: 200,
                  body: { data: 'PAYLOAD' },
                }),
              100
            );
          });
        },
        method: 'POST',
      });
      const rnl = new RelayNetworkLayer([
        retryMiddleware({
          fetchTimeout: 20,
          retryDelays: [1],
          logger: false,
        }),
      ]);

      expect.assertions(2);
      try {
        await mockReq(1).execute(rnl);
      } catch (e) {
        expect(e.toString()).toMatch('RelayNetworkLayer: reached request timeout in 20 ms');
      }
      const reqs = fetchMock.calls('/graphql');
      expect(reqs).toHaveLength(2);
    });

    it('should work forceRetry callback when request delayed', async () => {
      // First request will be fulfilled after 100ms delay
      // 2nd request and next without delays
      let attempt = 0;
      fetchMock.mock({
        matcher: '/graphql',
        response: () => {
          attempt++;
          return new Promise(resolve => {
            setTimeout(
              () =>
                resolve({
                  status: 200,
                  body: { data: 'PAYLOAD' },
                }),
              attempt === 1 ? 100 : 0
            );
          });
        },
        method: 'POST',
      });

      // will call force retry after 30 ms
      const forceRetry = jest.fn(runNow => {
        setTimeout(() => {
          runNow();
        }, 30);
      });

      const rnl = new RelayNetworkLayer([
        retryMiddleware({
          fetchTimeout: 10,
          retryDelays: () => 999,
          logger: false,
          forceRetry,
        }),
      ]);

      // make request
      const resPromise = mockReq(1).execute(rnl);
      await sleep(1);
      // should be sended first request (server will respond after 100 ms)
      expect(fetchMock.calls('/graphql')).toHaveLength(1);
      await sleep(10);
      // after 10 ms should be reached `fetchTimeout`
      // so middleware hang request
      // and starts 1000ms delayed period before making a new request
      // when delay period was started, should be called forceRetry method
      expect(forceRetry).toHaveBeenCalledTimes(1);
      // second arg of forceRetry call should be delay period in ms
      expect(forceRetry.mock.calls[0][1]).toBe(999);
      // on 30 ms will be called `runNow` function
      await sleep(50);
      // so we make second request before delay period will end
      expect(fetchMock.calls('/graphql')).toHaveLength(2);
      expect((await resPromise).data).toBe('PAYLOAD');
    });

    it('should call `onRetry` when request delayed', async () => {
      // First request will be fulfilled after 100ms delay
      // 2nd request and next without delays
      let attempt = 0;
      fetchMock.mock({
        matcher: '/graphql',
        response: () => {
          attempt++;
          return new Promise(resolve => {
            setTimeout(
              () =>
                resolve({
                  status: 200,
                  body: { data: 'PAYLOAD' },
                }),
              attempt === 1 ? 100 : 0
            );
          });
        },
        method: 'POST',
      });

      // will call force retry after 30 ms
      const onRetry = jest.fn(({ forceRetry }) => {
        setTimeout(() => {
          forceRetry();
        }, 30);
      });

      const rnl = new RelayNetworkLayer([
        retryMiddleware({
          fetchTimeout: 10,
          retryDelays: () => 999,
          logger: false,
          onRetry,
        }),
      ]);

      // make request
      const resPromise = mockReq(1).execute(rnl);
      await sleep(1);
      // should be sended first request (server will respond after 100 ms)
      expect(fetchMock.calls('/graphql')).toHaveLength(1);
      await sleep(10);
      // after 10 ms should be reached `fetchTimeout`
      // so middleware hang request
      // and starts 1000ms delayed period before making a new request
      // when delay period was started, should be called forceRetry method
      expect(onRetry).toHaveBeenCalledTimes(1);
      expect(onRetry.mock.calls[0][0]).toEqual({
        attempt: 1,
        delay: 999,
        forceRetry: expect.anything(),
        lastError: expect.objectContaining({
          message: expect.stringContaining('response timeout'),
        }),
      });
      // on 30 ms will be called `forceRetry` function
      await sleep(50);
      // so we make second request before delay period will end
      expect(fetchMock.calls('/graphql')).toHaveLength(2);
      expect((await resPromise).data).toBe('PAYLOAD');
    });

    it('should call `onRetry` and reject request if returned `false`', async () => {
      // First request will be fulfilled after 100ms delay
      // 2nd request and next without delays
      let attempt = 0;
      fetchMock.mock({
        matcher: '/graphql',
        response: () => {
          attempt++;
          return new Promise(resolve => {
            setTimeout(
              () =>
                resolve({
                  status: 200,
                  body: { data: 'PAYLOAD' },
                }),
              attempt === 1 ? 100 : 0
            );
          });
        },
        method: 'POST',
      });

      // will call force retry after 30 ms
      const onRetry = jest.fn(() => {
        return false;
      });

      const rnl = new RelayNetworkLayer([
        retryMiddleware({
          fetchTimeout: 10,
          retryDelays: () => 999,
          logger: false,
          onRetry,
        }),
      ]);

      // make request
      const resPromise = mockReq(1).execute(rnl);
      await sleep(1);
      // should be sended first request (server will respond after 100 ms)
      expect(fetchMock.calls('/graphql')).toHaveLength(1);
      await sleep(10);
      // after 10 ms should be reached `fetchTimeout`
      // so middleware hang request
      // and starts 1000ms delayed period before making a new request
      // when delay period was started, should be called forceRetry method
      expect(onRetry).toHaveBeenCalledTimes(1);
      expect(onRetry.mock.calls[0][0]).toEqual({
        attempt: 1,
        delay: 999,
        forceRetry: expect.anything(),
        lastError: expect.objectContaining({
          message: expect.stringContaining('response timeout'),
        }),
      });

      await resPromise.catch(e => {
        expect(e).toMatchObject({
          message: expect.stringContaining('interrupted in onRetry() callback'),
        });
      });

      // we should not make second request
      expect(fetchMock.calls('/graphql')).toHaveLength(1);
    });
  });
});
