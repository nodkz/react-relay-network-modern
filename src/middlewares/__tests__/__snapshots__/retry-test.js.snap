// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middlewares/retry middleware should make retries 1`] = `
Array [
  Array [
    "/graphql",
    Object {
      "body": "{\\"id\\":\\"1\\",\\"query\\":\\"\\",\\"variables\\":{}}",
      "headers": Object {
        "Accept": "*/*",
        "Content-Type": "application/json",
      },
      "method": "POST",
      "signal": AbortSignal {},
    },
  ],
  Array [
    "/graphql",
    Object {
      "body": "{\\"id\\":\\"1\\",\\"query\\":\\"\\",\\"variables\\":{}}",
      "headers": Object {
        "Accept": "*/*",
        "Content-Type": "application/json",
      },
      "method": "POST",
      "signal": AbortSignal {},
    },
  ],
  Array [
    "/graphql",
    Object {
      "body": "{\\"id\\":\\"1\\",\\"query\\":\\"\\",\\"variables\\":{}}",
      "headers": Object {
        "Accept": "*/*",
        "Content-Type": "application/json",
      },
      "method": "POST",
      "signal": AbortSignal {},
    },
  ],
]
`;

exports[`middlewares/retry middleware should retry request on timeout 1`] = `
Array [
  Array [
    "/graphql",
    Object {
      "body": "{\\"id\\":\\"1\\",\\"query\\":\\"\\",\\"variables\\":{}}",
      "headers": Object {
        "Accept": "*/*",
        "Content-Type": "application/json",
      },
      "method": "POST",
      "signal": AbortSignal {},
    },
  ],
  Array [
    "/graphql",
    Object {
      "body": "{\\"id\\":\\"1\\",\\"query\\":\\"\\",\\"variables\\":{}}",
      "headers": Object {
        "Accept": "*/*",
        "Content-Type": "application/json",
      },
      "method": "POST",
      "signal": AbortSignal {},
    },
  ],
  Array [
    "/graphql",
    Object {
      "body": "{\\"id\\":\\"1\\",\\"query\\":\\"\\",\\"variables\\":{}}",
      "headers": Object {
        "Accept": "*/*",
        "Content-Type": "application/json",
      },
      "method": "POST",
      "signal": AbortSignal {},
    },
  ],
]
`;
